### ISSUE TITLE
CPU Family/Model is used as Ext.Family/Model

### ISSUE THREAD

**Issue by Umio-Yasuno** [anrieff/libcpuid#150]
CPUID Family is expressed as *(Base.Family + Ext.Family)* ,
CPUID Model is expressed as *(Base.Model + Ext.Model)* .

[Intel® 64 and IA-32 Architectures Developer's Manual: Vol. 2A](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-2a-manual.html) (Page 307)
[PPR for AMD Family 17h Model 60h A1](https://www.amd.com/en/support/tech-docs?keyword=PPR) (Page 59)

However, in libcpuid, *(Base.Family)* store in `family` , *(Base.Family + Ext.Family)* store in `ext_family` . About Model is the same.

<https://github.com/anrieff/libcpuid/blob/52c5f505cff57266b32aa8eb95eb7c7fc47db94b/libcpuid/cpuid_main.c#L314>

I think of *Base, Extended, and Coupled Family/Model* should be distinguished.

Related: <https://github.com/X0rg/CPU-X/issues/185>

**Comment by Umio-Yasuno** [comment]
This is a fundamental issue, and any attempt to fix it will inevitably result in destructive changes for the library.

**Comment by anrieff** [comment]
Yes, I agree it is a fundamental issue, but more of a naming one. What libcpuid calls "extended family" and "extended model" are what is described in the docs as "Display family" and "Display model".

I'm not convinced though that the true Ext. Family / Ext. Model fields (as described in this issue and the Intel/AMD docs) are very useful in isolation. Consider them bit extensions of the Base Family / Model fields. This kind of thing happens quite often in binary data formats (typically files / headers, but the same principle applies to processor designs as well), when the original design was too restricted and too little bits were allotted for something that evolved, and eventually the design ran out of bits. In Pentium I / Pentium II the CPUID EAX=1 leaf only had Family ID and Model ID, each with just 4 bits. This restricted the future growth to just 16 families of 16 models each, which might have seemed plenty back then. Eventually the companies wanted to have much more models of the same family, and more family space as well. So they picked two unused portions of the CPUID EAX=1 leaf and added two more bit fields (called Ext. Family and Ext. Model) that, *in combination with the existing bits for Base Family/Model*, made the Family field essentially a 8-bit one, and the Model field essentially a 8-bit one. Only the Model was purely bit extended, the Family field uses an addition scheme, I guess for backwards compatibility with existing software back then.

All in all, I think this is a valid issue in pointing that the documentation is vague abouth what exactly is stored in `ext_family` and `ext_model` fields of `struct cpu_id_t`. I'll fix that.

**Comment by anrieff** [comment]
If you insist, you can always get the relevant fields manually, using `cpu_exec_cpuid(1, ...)`. The following code extracts all family/model related fields and prints them to stdout:

```
/*
 * Code snippet to extract all Model/Family-related fields from a CPU.
 */
#include <stdio.h>
#include <libcpuid.h>

int main(void)
{
	struct cpu_id_t cpu;
	uint32_t leaf1[4];
	
	cpu_identify(NULL, &cpu);
	cpu_exec_cpuid(1, leaf1);
	
	printf("CPU BaseFamily is 0x%x, ExtendedFamily is 0x%02x, effective (display) family is 0x%02x\n",
		   cpu.family, (leaf1[0] >> 20) & 0xff, cpu.ext_family);
	printf("CPU BaseModel is 0x%x, ExtendedModel is 0x%x, effective (display) model is 0x%02x\n",
		   cpu.model, (leaf1[0] >> 16) & 0xf, cpu.ext_model);
	return 0;
}
```

The output, on my Haswell mobile CPU, is:
```
CPU BaseFamily is 0x6, ExtendedFamily is 0x00, effective (display) family is 0x06
CPU BaseModel is 0xc, ExtendedModel is 0x3, effective (display) model is 0x3c
```

**Comment by anrieff** [comment]
Fixed in 27b0028.

**Comment by Umio-Yasuno** [comment]
Thank you for your reply and response. 
Sorry, I was shallow knowledge about the historical background of the Family/Model.

It was for [X0rg/CPU-X](https://github.com/X0rg/CPU-X) at first. Thanks to you, I think I can come up with another solution.
